contrib/vexecutor/vcheck.c

```diff
@@ -54,7 +61,274 @@ typedef struct VecFuncHashEntry
+        HeapTuple tuple;
+
+        //OpExpr mostly have two args, check the first one
+        argnode = linitial(op->args);
+        if(CheckVectorizedExpression(argnode, ctx))
+        	    return true;
+
+        ltype = ctx->retType;
+
+        //check the second one
+        argnode = lsecond(op->args);
+        if(CheckVectorizedExpression(argnode, ctx))
+        	    return true;
+
+        rtype = ctx->retType;
+
+        //check the return type
+        rettype = GetVtype(op->opresulttype);
+        if(InvalidOid == rettype)
+        	    return true;
+
+
+        //get the vectorized operator functions
+        //NOTE:we have no ParseState now, Give the NULL value is OK but not good...
+        tuple = oper(NULL, list_make1(makeString(get_opname(op->opno))),
+        				ltype, rtype, false, -1);
+        if(NULL == tuple)
+        	    return true;
+
+        voper = (Form_pg_operator)GETSTRUCT(tuple);
+        if(voper->oprresult != rettype)
+        	    return true;
+
+        if(ctx->replace)
+        {
+        	    op->opresulttype = rettype;
+        	    op->opfuncid = voper->oprcode;
+        }
+
+        ctx->retType = rettype;
+        return false;
+    }
+
+    //now, other nodes treat as can not be vectorized
+    return plan_tree_walker(node, CheckVectorizedExpression, ctx);;
+}
+
+/*
+ * check an plan node, all the expressions in it should be checked
+ * set the flag if an plan node can be vectorized
+ */
+static bool
+CheckPlanNodeWalker(PlannerInfo *root, Plan *plan)
+{
+	VectorizedContext ctx;
+
+    if(plan->vectorized)
+    	    return true;
+
+    ctx.replace =false;
+
+    ctx.retType = InvalidOid;
+    plan->vectorized = !plan_tree_walker((Node*)plan,
+    							 CheckVectorizedExpression,
+							 &ctx);
+
+
+    return false;
+}
+
+/*
+ * check the plan tree
+ */
+static Plan*
+CheckPlanVectorzied(PlannerInfo *root, Plan *plan)
+{
+    if(NULL == plan)
+        return plan;
+
+    CheckPlanVectorzied(root, plan->lefttree);
+    CheckPlanVectorzied(root, plan->righttree);
+    CheckPlanNodeWalker(root, plan);
+
+    return plan;
+}
+
+/*
+ * Replace the non-vectorirzed type to vectorized type
+ */
+static bool
+ReplacePlanNodeWalker(PlannerInfo *root, Plan *plan)
+{
+	VectorizedContext ctx;
+
+    if(!plan->vectorized)
+    		return false;
+
+    if(!HasVecExecOprator(nodeTag(plan)))
+    {
+    		plan->vectorized = false;
+    		return false;
+    }
+
+    ctx.replace =true;
+
+    ctx.retType = InvalidOid;
+    plan_tree_walker((Node*)plan,
+    					CheckVectorizedExpression,
+					&ctx);
+
 
+    return false;
+}
+
+/*
+ * check the plan tree
+ */
+static Plan*
+ReplacePlanVectorzied(PlannerInfo *root, Plan *plan)
+{
+    if(NULL == plan)
+        return plan;
+
+    ReplacePlanVectorzied(root, plan->lefttree);
+    ReplacePlanVectorzied(root, plan->righttree);
+    ReplacePlanNodeWalker(root, plan);
+
+    return plan;
+}
+
+Plan*
+CheckAndReplacePlanVectorized(PlannerInfo *root, Plan *plan)
```

Comment:
How to check initPlans of main Plan node?

Reply:
So far it is not supported, but it does not seem to affect the normal running.